csrc/quickreduce/quick_reduce.h

```diff
@@ -0,0 +1,224 @@
+  } else if (world_size == 4) {                                             \
+    using AllReduceKernel = AllReduceOneshot<T, 4>;                         \
+    hipLaunchKernelGGL((allreduce_prototype_oneshot<AllReduceKernel, T>),   \
+                       dim3(grid), dim3(kBlockOneShot), 0, stream, A, B, N, \
+                       rank, dbuffer_list, data_offset, flag_color);        \
+  } else if (world_size == 8) {                                             \
+    using AllReduceKernel = AllReduceOneshot<T, 8>;                         \
+    hipLaunchKernelGGL((allreduce_prototype_oneshot<AllReduceKernel, T>),   \
+                       dim3(grid), dim3(kBlockOneShot), 0, stream, A, B, N, \
+                       rank, dbuffer_list, data_offset, flag_color);        \
+  }
+
+#define TWOSHOT_DISPATCH(__codec)                                           \
+  if (world_size == 2) {                                                    \
+    using LineCodec = __codec<T, 2>;                                        \
+    using AllReduceKernel = AllReduceTwoshot<T, LineCodec>;                 \
+    hipLaunchKernelGGL((allreduce_prototype_twoshot<AllReduceKernel, T>),   \
+                       dim3(grid), dim3(kBlockTwoShot), 0, stream, A, B, N, \
+                       num_blocks, rank, dbuffer_list, data_offset,         \
+                       flag_color);                                         \
+  } else if (world_size == 4) {                                             \
+    using LineCodec = __codec<T, 4>;                                        \
+    using AllReduceKernel = AllReduceTwoshot<T, LineCodec>;                 \
+    hipLaunchKernelGGL((allreduce_prototype_twoshot<AllReduceKernel, T>),   \
+                       dim3(grid), dim3(kBlockTwoShot), 0, stream, A, B, N, \
+                       num_blocks, rank, dbuffer_list, data_offset,         \
+                       flag_color);                                         \
+  } else if (world_size == 8) {                                             \
+    using LineCodec = __codec<T, 8>;                                        \
+    using AllReduceKernel = AllReduceTwoshot<T, LineCodec>;                 \
+    hipLaunchKernelGGL((allreduce_prototype_twoshot<AllReduceKernel, T>),   \
+                       dim3(grid), dim3(kBlockTwoShot), 0, stream, A, B, N, \
+                       num_blocks, rank, dbuffer_list, data_offset,         \
+                       flag_color);                                         \
+  }
+
+struct DeviceComms {
+  // Workgroup scope = Tile = (256 threads x 16B x 8 atoms)
+  static long constexpr kTileSize = 256 * 16 * 8;
+
+  // Max problem size is 8GB (in bytes)
+  static long constexpr kMaxProblemSize = 8589934592;
+  static long constexpr kMaxTiles = kMaxProblemSize / kTileSize;
+
+  // Max TP-8
+  static int constexpr kMaxWorldSize = 8;
+
+  bool initialized = false;
+  int flag_color = 1;
```

Comment:
To make it robust in cases of rolling over:
```suggestion
  uint32_t flag_color = 1;
```